proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+
+In Haystack 1.x, `PromptNode` uses `InvocationLayer` to query different LLMs under a unified API. In that design, users do not need to know which invocation layer is used for the model they select, as `PromptNode` takes responsibility of selecting it.
+
+Such invocation layers can be ported to 2.0 as standalone components. In this way we will have one component for each LLM backed that we support.
+
+Each component should be named after the class of models it supports. For example we should have `GPT4`, `ChatGPT`, etc. For components whose name may be confusing, or that include models that are not LLMs, we can add a LLM suffix: like `FalconLLM`, `CohereLLM`, `HuggingFaceLLM`.
+
+Note that having separate components for each LLM makes easy to deprecate them when we realize they are dropping out of favor or become severely outdated. It also makes very easy for external contributors to make their own external components to support rarer LLMs, without having to add them to Haystack’s core.
+
+All these LLM clients will have a near-identical I/O:
+
+```python
+@component
+class ChatGPT:
+
+    @component.output_types(replies=List[List[str]])
+    def run(self, prompts: List[str], ... chatgpt specific params...):
+        ...
+        return {'replies': [...]}
+```
+
+Note how the component takes a list of prompts and LLM parameters only, but no variables nor templates, and returns only strings. This is because input rendering and output parsing are delegated to separate components, which description follows.
+
+### Returning metadata
+
+In the example above we made the LLM return only a list of replies, as strings. However, in order to be able to parse the output into meaningful objects (see “Output parsing”) we may need additional metadata from these clients.
+
+1. Do we already have any such situation?
+2. Can we foresee any other?
+
+If the answer to any of the above is yes, a simple, maybe temporary solution would be to add a second output, called for example `'replies_meta'` . Any component that need such meta to parse the output would then request this second output along with the first and zip the two lists together to reconstruct the original output of the LLM.
+
+### Returning streams
+
+In some cases users may want to see the output of the LLM as it’s being generated, so as a stream that prints word by word in the terminal.
+
+To achieve such effect, we should make LLM components return a more generic type `replies: List[Iterator[str]]`. In this way, users will have the ability of returning a generator or iterator in place of a regular string, and therefore allow the receiving components to unroll these replies as they wish. Components that need the full reply will unroll the iterator internally (like `RepliesToAnswersConverter`), while if the generator goes in the output, the users can unroll it in a loop printing the content word by word. We can also have specialized components to visualize the stream within a pipeline: something like:
```

Comment:
Thread: https://www.notion.so/deepsetai/Notes-about-streaming-the-LLM-outputs-7ca49b892237449fad59bd0e7cc03e69